test/cpp/end2end/client_fork_test.cc

```diff
@@ -0,0 +1,173 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#ifndef GRPC_ENABLE_FORK_SUPPORT
+// No-op for builds without fork support.
+int main(int /* argc */, char** /* argv */) { return 0; }
+#else  // GRPC_ENABLE_FORK_SUPPORT
+
+#include <signal.h>
+
+#include <gtest/gtest.h>
+
+#include "absl/strings/str_cat.h"
+
+#include <grpc/fork.h>
+#include <grpc/grpc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/time.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+#include <grpcpp/server_context.h>
+
+#include "src/core/lib/gprpp/fork.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/core/util/port.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/util/test_config.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+using ::grpc::testing::EchoRequest;
+using ::grpc::testing::EchoResponse;
+using ::grpc::testing::EchoTestService;
+
+class ServiceImpl final : public EchoTestService::Service {
+  Status BidiStream(
+      ServerContext* /*context*/,
+      ServerReaderWriter<EchoResponse, EchoRequest>* stream) override {
+    EchoRequest request;
+    EchoResponse response;
+    while (stream->Read(&request)) {
+      gpr_log(GPR_INFO, "recv msg %s", request.message().c_str());
+      response.set_message(request.message());
+      stream->Write(response);
+      gpr_log(GPR_INFO, "wrote msg %s", response.message().c_str());
+    }
+    return Status::OK;
+  }
+};
+
+std::unique_ptr<EchoTestService::Stub> MakeStub(const std::string& addr) {
+  return EchoTestService::NewStub(
+      grpc::CreateChannel(addr, InsecureChannelCredentials()));
+}
+
+TEST(ClientForkTest, ClientCallsBeforeAndAfterForkSucceed) {
+  grpc_core::Fork::Enable(true);
+
+  int port = grpc_pick_unused_port_or_die();
+  std::string addr = absl::StrCat("localhost:", port);
+
+  pid_t server_pid = fork();
+  switch (server_pid) {
+    case -1:  // fork failed
+      GTEST_FAIL() << "failure forking";
+    case 0:  // post-fork child
+    {
+      ServiceImpl impl;
+      grpc::ServerBuilder builder;
+      builder.AddListeningPort(addr, grpc::InsecureServerCredentials());
+      builder.RegisterService(&impl);
+      std::unique_ptr<Server> server(builder.BuildAndStart());
+      server->Wait();
+      return;
+    }
+    default:  // post-fork parent
+      break;
+  }
+
+  // Do a round trip before we fork.
+  // NOTE: without this scope, test running with the epoll1 poller will fail.
```

Comment:
Note that without this scope, test running with epoll1 poller will fail. Also note that the EE poll poller hanging issue only happens when this scope is removed.

Reply:
Is this still true?

Reply:
Yes, this is still true. I'm not sure what causes the failure though.